src/components/Resource/Forms/TrainingSelectFile.tsx

```diff
@@ -0,0 +1,311 @@
+        title: 'Failed to create file.',
+        message: error.message,
+      });
+    },
+  });
+
+  const upsertVersionMutation = trpc.modelVersion.upsert.useMutation({
+    onSuccess: async (vData) => {
+      // TODO [bw] ideally, we would simply update the proper values rather than invalidate to skip the loading step
+      await queryUtils.modelVersion.getById.invalidate({ id: vData.id });
+      await queryUtils.model.getById.invalidate({ id: model?.id });
+      await queryUtils.model.getMyTrainingModels.invalidate();
+      setAwaitInvalidate(false);
+      onNextClick();
+    },
+    onError: (error) => {
+      setAwaitInvalidate(false);
+      showErrorNotification({
+        error: new Error(error.message),
+        title: 'Failed to save model version',
+      });
+    },
+  });
+
+  const handleSubmit = async () => {
+    if (!model || !modelVersion) {
+      showErrorNotification({
+        error: new Error('Missing model data. Please try again.'),
+      });
+      return;
+    }
+    if (!selectedFile || !selectedFile.length) {
+      showErrorNotification({
+        error: new Error('Please select a file to be used.'),
+      });
+      return;
+    }
+
+    setAwaitInvalidate(true);
+
+    showNotification({
+      id: notificationId,
+      loading: true,
+      autoClose: false,
+      disallowClose: true,
+      title: 'Applying selected model file...',
+      message: '',
+    });
+
+    const result = await fetch(selectedFile);
+    if (!result.ok) {
+      setAwaitInvalidate(false);
+      updateNotification({
+        id: notificationId,
+        icon: <IconX size={18} />,
+        color: 'red',
+        title: 'Failed to create file.',
+        message: 'Please try again.',
+      });
+      console.log(result);
+      return;
+    }
+    const blob = await result.blob();
+    const blobFile = new File([blob], `${selectedFile.split('/').pop()}`, {
+      type: blob.type,
+    });
+
+    await upload(
+      {
+        file: blobFile,
+        type: UploadType.Model,
+        meta: {
+          versionId: modelVersion.id,
+          // TODO more
+        },
+      },
+      async ({ meta, size, ...result }) => {
+        const { versionId, type, uuid, ...metadata } = meta as {
+          versionId: number;
+          type: ModelFileType;
+          uuid: string;
+        };
+        console.log(type);
+        createFileMutation.mutate({
+          ...result,
+          sizeKB: bytesToKB(size),
+          modelVersionId: versionId,
+          type: 'Model',
+          metadata,
+        });
+      }
+    );
+  };
+
+  // you should only be able to get to this screen after having created a model, version, and uploading a training set
+  if (!model || !modelVersion || !modelFile) {
+    return <NotFound />;
+  }
+
+  const epochs = modelFile.metadata.trainingResults?.epochs.sort(function (a, b) {
+    const x = a.epoch_number;
+    const y = b.epoch_number;
+    return x > y ? -1 : x < y ? 1 : 0;
+  });
```

Comment:
Just nitpicking to save a few lines 😛 
```suggestion
  const epochs = modelFile.metadata.trainingResults?.epochs.sort((a, b) => b - a);
```
Maybe consider cloning the epochs array cause `.sort` mutates the original reference as well

Reply:
good idea, but I think you mean this:

```
const epochs = [...(modelFile.metadata.trainingResults?.epochs || [])].sort(
    (a, b) => b.epoch_number - a.epoch_number
  );
```
with the epoch_number field being used